### ISSUE TITLE
Preserve comments

### ISSUE THREAD

**Issue by fvictorio** [prettier-solidity/prettier-plugin-solidity#32]
Comments are being dropped out right now. They should at least be preserved; properly formatting them could be another issue (which will probably need some definition: for example, should long comment lines be split? What does prettier do in js?).

**Comment by mattiaerre** [comment]
yes, we should definitely preserve comments; no need to properly formatting but just keep them as they are (w/ the exception of removing unneeded spaces at the end of lines); are you able to attempt a PR for this issue?

**Comment by fvictorio** [comment]
I'll give it a try! Please assign it to me.

**Comment by fvictorio** [comment]
I tried to do this but I had a problem with the grammar. Right now the comments are sent to the hidden channel:

https://github.com/solidityj/solidity-antlr4/blob/f4b1b9a98348f5ec0dc525854a9ad7aee2a47f64/Solidity.g4#L454-L458

There seems to be a way to access them, but I don't know enough antlr to do it (and this may not be available as a user of the solidity-parser module).

A possible solution would be to expose the comments as nodes in the AST. I tried to modify the grammar so that they aren't sent to HIDDEN, but this isn't enough: some modifications to the AST construction may be needed too. Besides, this may be a breaking change for other users of the grammar and the parser.

@federicobond do you think this is a change worth doing? For comparison, `babylon` does include the comments as nodes in the generated tree.

**Comment by federicobond** [comment]
It's difficult to capture them in the AST because token can appear inside other AST nodes, like this:

```
public /* private */ uint a = 1;
```

The ultimate solution is to provide a Concrete Syntax Tree that preserves all these. Alternatively, extracting comments using a Regex or a simple state machine like [this](https://github.com/federicobond/solidity-extract-imports) is probably a better option for now.

**Comment by mattiaerre** [comment]
that's an interesting approach @federicobond; so you are suggesting to decorate the parser to extract comments and add them to the returned AST?

https://github.com/prettier-solidity/prettier-plugin-solidity/blob/master/src/parser.js#L1-L6

**Comment by federicobond** [comment]
Injecting them in the AST is one way, but you could even keep them separately and consume the comment spans from a list as you visit and print the tree. Not sure if the bookkeeping will become too complicated but seems doable.

**Comment by mattiaerre** [comment]
thanks @federicobond for this, I also was looking at an extract functionality and I've found this library:

https://github.com/jonschlinkert/extract-comments#usage

the real question is not how to extract comment but rather how to put them back into the formatted code 😅

**Comment by fvictorio** [comment]
Apparently you have to add a `comments` property to the root of the AST. Something like:

```
const parse = text => {
  const result = parser.parse(text, { loc: true, range: true })

  result.comments = extractComments(text)

  return result
};
```

That didn't work for me, though. Now prettier throws a: `Error: Comment "/* set the message */" was not printed. Please report this error!`.

@vjeux, sorry for bothering you, but I saw that you helped the folks of `prettier-python` and `prettier-java` with similar issues. I extracted the comments of the file and added an array to the root of the AST with this format:

```
{
  type: String,
  loc: { start: { line, column }, end: { ... } },
  value: String
}
```

But I'm getting that error with some simple example. Is there something else I have to add to the comments array?

**Comment by vjeux** [comment]
You also need to implement printComments function: https://github.com/prettier/prettier/blob/master/src/language-graphql/printer-graphql.js#L630-L641

**Comment by vjeux** [comment]
This is awesome that you're working on a pretty printer for solidity btw!

**Comment by mattiaerre** [comment]
thanks @vjeux yes we've noticed that; there is also this PR open (w/ a lot of copy-paste from the prettier js lang) if you can have a look #40 😊

**Comment by mattiaerre** [comment]
fixed by #40
